oak_sign/src/lib.rs

```diff
@@ -31,44 +31,59 @@ pub const PUBLIC_KEY_TAG: &str = "PUBLIC KEY";
 pub const SIGNATURE_TAG: &str = "SIGNATURE";
 pub const HASH_TAG: &str = "HASH";
 
-/// Convenience struct that encapsulates `Ed25519KeyPair`.
-/// Named `KeyBundle` since it uses the `ring::signature::KeyPair` trait.
-pub struct KeyBundle {
-    private_key: Vec<u8>,
-    key_pair: Ed25519KeyPair,
+/// Convenience struct that encapsulates `ring::signature::Ed25519KeyPair`.
+pub struct KeyPair {
+    /// PKCS#8 v2 encoded private key and public key.
+    /// https://tools.ietf.org/html/rfc5958
+    ///
+    /// The encoded version is kept because a `parsed` key pair doesn't contain a method for
+    /// extracting private key bytes.
+    /// https://briansmith.org/rustdoc/ring/signature/struct.Ed25519KeyPair.html
+    encoded_key_pair: Vec<u8>,
+    /// Parsed PKCS#8 v2 key pair representation.
+    parsed_key_pair: Ed25519KeyPair,
 }
 
-impl KeyBundle {
+impl Clone for KeyPair {
+    fn clone(&self) -> Self {
+        KeyPair::parse(&self.encoded_key_pair).expect("Couldn't clone key pair")
+    }
+}
+
+impl KeyPair {
     /// Generates a Ed25519 key pair.
-    pub fn generate() -> anyhow::Result<KeyBundle> {
+    pub fn generate() -> anyhow::Result<KeyPair> {
         let rng = rand::SystemRandom::new();
         let private_key = Ed25519KeyPair::generate_pkcs8(&rng)
             .ok()
             .context("Couldn't generate key pair")?;
-        KeyBundle::parse(private_key.as_ref())
+        KeyPair::parse(private_key.as_ref())
     }
 
     /// Parses a Ed25519 key pair from bytes.
-    pub fn parse(private_key: &[u8]) -> anyhow::Result<KeyBundle> {
-        let key_pair = Ed25519KeyPair::from_pkcs8(private_key)
+    pub fn parse(private_key: &[u8]) -> anyhow::Result<KeyPair> {
+        let parsed_key_pair = Ed25519KeyPair::from_pkcs8(private_key)
             .ok()
             .context("Couldn't parse generated key pair")?;
         Ok(Self {
-            private_key: private_key.to_vec(),
-            key_pair,
+            encoded_key_pair: private_key.to_vec(),
+            parsed_key_pair,
         })
     }
 
     pub fn private_key(&self) -> Vec<u8> {
```

Comment:
Is this method what `parse` reads back? Could you add a comment and ideally also some tests that encode and decode a key?

And also please add some other tests of various functionality, since this crate is more and more used for other purposes too.

Reply:
Done
Also added tests